app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt

```diff
@@ -0,0 +1,126 @@
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.logging.speziLogger
+import edu.stanford.spezi.module.onboarding.invitation.await
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import org.hl7.fhir.r4.model.Observation
+import javax.inject.Inject
+
+internal class ObservationRepository @Inject constructor(
+    private val firestore: FirebaseFirestore,
+    private val firebaseAuth: FirebaseAuth,
+    private val observationToRecordMapper: ObservationToRecordMapper,
+    @Dispatching.IO private val ioDispatcher: CoroutineDispatcher,
+) {
+    private val logger by speziLogger()
+
+    private val heartRateCollection =
+        ObservationCollection("heartRateObservations", Loinc.HeartRate.CODE)
+
+    private val bodyWeightObservation =
+        ObservationCollection("bodyWeightObservations", Loinc.Weight.CODE)
+
+    private val bloodPressureCollection =
+        ObservationCollection("bloodPressureObservations", Loinc.BloodPressure.CODE)
+
+    private val jsonParser by lazy {
+        FhirContext.forCached(FhirVersionEnum.R4).newJsonParser()
+    }
+
+    private val gson by lazy {
+        Gson()
+    }
+
+    suspend fun saveObservations(observations: List<Observation>) {
+        withContext(ioDispatcher) {
+            runCatching {
+                val uid = firebaseAuth.currentUser?.uid
+                    ?: throw IllegalStateException("User not authenticated")
+                val batch = firestore.batch()
+
+                observations.forEach { observation ->
+                    val collectionName = getCollectionNameForObservation(observation)
+                    val json = jsonParser.encodeResourceToString(observation)
+                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
+                    val data = gson.fromJson<Map<String, Any>>(json, mapType)
+
+                    val docRef = firestore.collection("users/$uid/$collectionName").document()
+                    batch.set(docRef, data)
+                }
+
+                batch.commit().await()
+            }.onFailure {
+                logger.e(it) { "Error while saving observations" }
+            }
+        }
+    }
+
+    private fun getCollectionNameForObservation(observation: Observation): String {
+        return when {
+            observation.code.coding.any { it.code == bodyWeightObservation.code } -> bodyWeightObservation.name
+            observation.code.coding.any { it.code == bloodPressureCollection.code } -> bloodPressureCollection.name
+            observation.code.coding.any { it.code == heartRateCollection.code } -> heartRateCollection.name
+            else -> {
+                logger.w { "Unknown observation code" }
+                throw UnsupportedOperationException("Unknown observation code")
+            }
+        }
+    }
+
+    private suspend fun <T> getObservation(collection: ObservationCollection): Result<T?> {
```

Comment:
Might make sense to rather rely on a snapshot listener + the overall list of collections and get the first element from there rather than multiple Firebase queries.